### PR TITLE
Allow user to sort by status

### PR DESCRIPTION
**Pull request by hiqua**
Allow user to sort by priority, i.e. in this order
* not done (NO)
* "done" because done the previous days and the habit requires less than once per day (YES_AUTO), but not ticked today
* done (YES_MANUAL)

Partially solves #565 and #579.

[Screenshot_20201115_153752](https://user-images.githubusercontent.com/3812042/99187747-a42dc380-2758-11eb-88f7-2212fd96cbed.png)

### CODE REVIEW

**Comment by iSoron** [comment]
Thank you for the pull request, @hiqua. After testing this, I think it would actually make more sense for us to add a checkbox "Move completed habits to the bottom". In this way, the user could still sort by score, for example, but have the items move out of the way when completed. I think it would be more complicated to implement than your current solution, but much better from a user perspective.

Would you like to work on this?

**Comment by hiqua** [comment]
@iSoron thanks for the review. Another possibility would be to remember the previous sorting method along with the current one, and if the current one returns an equality, compare with the previous one. To have the configuration you mention, a user would then need to:
* select score sorting
* select status sorting

From the user interface, we could just keep things like they currently are. The feature would thus be a bit hidden, but still discoverable enough.

If you don't like this suggestion I'll implement yours on a new PR.

**Comment by iSoron** [comment]
> Another possibility would be to remember the previous sorting method along with the current one, and if the current one returns an equality, compare with the previous one.

I like this solution. It would also work with other pairs of sorting methods (e.g. "color" then "score").

**Comment by hiqua** [comment]
Sorting by color then priority:
Screenshot_20201117_232541

Sorting by name then priority:
Screenshot_20201117_232604

**Comment by iSoron** [comment]
Thanks for the updates, @hiqua. A few comments:

1. Could you rename "sort by priority" to "sort by status"? Status sounds more clear to me. "Priority" sounds like something that the user could assign to each habit.
2. It seems like we are not persisting the secondary sorting method, only the primary one, so the ordering changes after the app restarts.
3. Most comparators we previously had (color, score, position) are hardcoded to fall back to comparison by name in case of ties. I think it makes sense to let them fall back to the secondary sorting method instead. This would allow some interesting things such as sorting by color, but moving completed habits to the bottom of the color category when completed ("sort by status" then "sort by color").

**Comment by hiqua** [comment]
1. done
2. done (I think), but I'm gonna need a review for that, it was more code than I expected. I should probably write more tests as well, but I'll wait for you to tell me whether you're somewhat satisfied with the current implementation first 
3. done

File: android/uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/views/HabitCardListAdapter.java [hunk 1]
```diff
@@ -83,6 +83,7 @@ public HabitCardListAdapter(@NonNull HabitCardListCache cache,
         cache.setListener(this);
         cache.setCheckmarkCount(
             ListHabitsRootViewKt.MAX_CHECKMARK_COUNT);
+        cache.setPreviousOrder(preferences.getDefaultPreviousOrder());
         cache.setOrder(preferences.getDefaultOrder());
```
Comment (iSoron) [inline android/uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/views/HabitCardListAdapter.java:87]: Could you please rename `order` and `previousOrder` to `primaryOrder` and `secondaryOrder` everywhere? I think this is more clear. For example, these two call would become `cache.setPrimaryOrder(...)` and `cache.setSecondaryOrder(...)`

File: android/uhabits-core/src/test/java/org/isoron/uhabits/core/io/HabitsCSVExporterTest.java [hunk 2]
```diff
@@ -68,11 +68,11 @@ public void testExportCSV() throws IOException
...
-        assertPathExists("001 Wake up early");
-        assertPathExists("001 Wake up early/Checkmarks.csv");
-        assertPathExists("001 Wake up early/Scores.csv");
-        assertPathExists("002 Meditate/Checkmarks.csv");
-        assertPathExists("002 Meditate/Scores.csv");
+        assertPathExists("001 Meditate/Checkmarks.csv");
+        assertPathExists("001 Meditate/Scores.csv");
+        assertPathExists("002 Wake up early");
+        assertPathExists("002 Wake up early/Checkmarks.csv");
+        assertPathExists("002 Wake up early/Scores.csv");
```
Comment (iSoron) [inline android/uhabits-core/src/test/java/org/isoron/uhabits/core/io/HabitsCSVExporterTest.java:75]: Why has this changed? It seems unrelated to the feature.

File: android/uhabits-core/src/test/java/org/isoron/uhabits/core/models/HabitListTest.java [hunk 3]
```diff
@@ -131,56 +130,73 @@ public void testOrdering()
...
         h4.setColor(1);
         h4.setPosition(2);
+        HabitList list;
 
-        list.add(h3);
-        list.add(h1);
-        list.add(h4);
-        list.add(h2);
-
+        list = testOrderingSetup(h1, h2, h3, h4);
```
Comment (iSoron) [inline android/uhabits-core/src/test/java/org/isoron/uhabits/core/models/HabitListTest.java:135]: Why do we need to recreate the list repeatedly? Does the previous code not work anymore?

File: android/uhabits-core/src/test/java/org/isoron/uhabits/core/io/ImportTest.java [hunk 4]
```diff
@@ -81,7 +81,7 @@ public void testRewireDB() throws IOException
 
         assertThat(habitList.size(), equalTo(3));
 
-        Habit habit = habitList.getByPosition(0);
+        Habit habit = habitList.getByPosition(1);
```
Comment (iSoron) [inline android/uhabits-core/src/test/java/org/isoron/uhabits/core/io/ImportTest.java:84]: Same as the previous comment.

**Review by iSoron — CHANGES_REQUESTED** [review]
Thank you for the changes. Please see some comments below.

Regarding tests, I would suggest better testing the logic within `onSortToggleBy`. For example, we could call `onSortByName` and `onSortByStatus` a few times and make sure that all the `setOrder` and `setPreviousOrder` call are correct.

File: android/uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/views/HabitCardListAdapter.java (hunk 1 above)
Comment (hiqua) [reply android/uhabits-android/src/main/java/org/isoron/uhabits/activities/habits/list/views/HabitCardListAdapter.java]: Done.

File: android/uhabits-core/src/test/java/org/isoron/uhabits/core/io/HabitsCSVExporterTest.java (hunk 2 above)
Comment (hiqua) [reply android/uhabits-core/src/test/java/org/isoron/uhabits/core/io/HabitsCSVExporterTest.java:75]: The default secondaryOrder is by name: https://github.com/iSoron/uhabits/pull/660/files#diff-36c07e8fa292800a1e6f010c52b7d81b33e2f1bc3514577d0c60df3d50959a86R42

Before my changes, this test was using the position sort. Given that these habits have the same position (`0`), they were left as they are inserted, so sorted by id. The changes reflect the facts that the habits are now sorted by name.

File: android/uhabits-core/src/test/java/org/isoron/uhabits/core/io/ImportTest.java (hunk 4 above)
Comment (hiqua) [reply android/uhabits-core/src/test/java/org/isoron/uhabits/core/io/ImportTest.java:84]: Also same answer, the default secondary ordering is by name, so the insertion ordering is no longer there.

**Comment by hiqua** [comment]
Test implemented, nothing left to do from my side.
